proto/pulumi/resource.proto

```diff
@@ -324,11 +325,14 @@ message ResourceHookRequest {
     google.protobuf.Struct old_inputs = 6; // the optional checked old inputs of the resource.
     google.protobuf.Struct new_outputs = 7; // the optional new outputs of the resource.
     google.protobuf.Struct old_outputs = 8; // the optional old outputs of the resource.
+    optional string failed_operation = 9; // for on_error, the operation that failed (create, update, or delete).
+    repeated string errors = 10; // for on_error, the errors that have been seen so far.
 }
 
 // ResourceHookResponse is the response object for resource hook callbacks in CallbackInvokeResponse.
 message ResourceHookResponse {
     string error = 1; // an optional error message to return from the hook.
+    bool retry = 2; // for on_error, whether we should retry the operation.
```

Comment:
I am slightly uneasy about this. It feels like maybe this should be a new type of callback rather than literally just a hook. I don't like the protocol now says any hook can return a retry flag but the engine just knows to ignore that for anything but a retry hook. Similarly for the errors list being a parameter to every hook but just only being filled in for error hooks, it was already a bit odd that we had so many optional properties (e.g. create never having old inputs/outputs), but they were similar enough it felt ok. This feels like its really pushing that envelope.

Reply:
Hmmm right now in `RegisterResourceHook` we create a function via `wrapResourceHookCallback` that we can call when it's time to invoke the hook later, similar to `wrapTransformCallback` for transforms.

I guess we could store the callback target & token instead of this function, and then when we get to calling the hook, we will know if it's an error or normal hook, and so we can pick the right request/response protobuf messages to serialize to/from.

Reply:
Alternatively, have a `RegisterErrorHook` instead.